src/lib/sticky-header/sticky-header-dir.ts

```diff
@@ -0,0 +1,424 @@
+})
+export class CdkStickyHeader implements OnDestroy, AfterViewInit {
+
+  /**
+   * Set the sticky-header's z-index as 10 in default. Make it as an input
+   * variable to make user be able to customize the zIndex when
+   * the sticky-header's zIndex is not the largest in current page.
+   * Because if the sticky-header's zIndex is not the largest in current page,
+   * it may be sheltered by other element when being stuck.
+   */
+  @Input('cdkStickyHeaderZIndex') zIndex: number = 10;
+
+  isStuck: boolean = false;
+  /**
+   * isStickyPositionSupported == true, means current browser support 'position: sticky'
+   * or 'position: -webkit-sticky'.
+   * isStickyPositionSupported == false, means current browser does not support sticky
+   * positioning and need to use the original implementation to get the sticky effect.
+   * @type {boolean}
+   */
+  isStickyPositionSupported: boolean = true;
+
+
+  /** The element with the 'cdkStickyHeader' tag. */
+  element: HTMLElement;
+  /** The upper container element with the 'cdkStickyRegion' tag. */
+  stickyParent: HTMLElement | null;
+  /** The upper scrollable container. */
+  upperScrollableContainer: HTMLElement;
+
+  /**
+   * The padding of the sticky-header. Put it here to avoid calling getComputedStyle()
+   * too many times.
+   */
+  padding: string;
+
+  /**
+   * The height of the sticky-region. Put it here to avoid calling getComputedStyle()
+   * too many times.
+   */
+  stickyRegionHeight: number;
+  /**
+   * The original css of the sticky element, used to reset the sticky element
+   * when it is being unstuck
+   */
+  originalCss: any;
+  /**
+   * 'getBoundingClientRect().top' of CdkStickyRegion of current sticky header.
+   * It is used with '_scrollFinish' to judge whether the current header
+   * need to be stuck.
+   */
+  private _containerStart: number;
+  /**
+   * It is 'The bottom of CdkStickyRegion of current sticky header - the height
+   * of current header', which is used with '_containerStart' to judge whether the current header
+   * need to be stuck.
+   */
+  private _scrollFinish: number;
+  /** The width of the sticky-header when it is stuck. */
+  private _scrollingWidth: number;
+
+  private _onScrollSubscription: Subscription;
+
+  private _onTouchSubscription: Subscription;
+
+  private _onResizeSubscription: Subscription;
+
+  constructor(element: ElementRef,
+              scrollable: Scrollable,
+              @Optional() public parentRegion: CdkStickyRegion) {
+    this.element = element.nativeElement;
+    this.upperScrollableContainer = scrollable.getElementRef().nativeElement;
+    this.setStrategyAccordingToCompatibility();
+  }
+
+  ngAfterViewInit(): void {
+    if (!this.isStickyPositionSupported) {
+      this.stickyParent = this.parentRegion != null ?
+        this.parentRegion._elementRef.nativeElement : this.element.parentElement;
+      this.originalCss = this.generateCssStyle(
+        this.getCssValue(this.element, 'zIndex'),
```

Comment:
This is still calling `getComputedStyle` over and over again.

Also, for top, left, right, bottom, height, and width you probably want the results of `getBoundingClientRect` instead of `getComputedStyle`